docs/developer/proposal/EP-002-MSR-Fallback-Power-Meter.md

```diff
@@ -0,0 +1,112 @@
+# EP-002: MSR Fallback for CPU Power Meter
+
+* **Status**: Draft
+* **Author**: Sunil Thaha
+* **Created**: 2025-08-13
+
+## Problem
+
+Kepler fails when `/sys/class/powercap/intel-rapl` is unavailable (disabled
+kernels, restricted containers).
+
+## Goals
+
+* Implement MSR-based RAPL reading as automatic fallback when powercap is
+  unavailable
+* Maintain existing CPUPowerMeter interface compatibility
+* Provide configurable control over fallback behavior for security-conscious
+  deployments
+
+## Solution
+
+Add MSR as fallback to read RAPL energy counters directly from CPU registers.
+
+## Proposed Solution
+
+```mermaid
+graph TB
+    CPUPowerMeter[CPUPowerMeter Interface]
+    raplPowerMeter[raplPowerMeter<br>Enhanced with MSR]
+    powercapReader[powercapReader<br>Primary]
+    msrReader[msrReader<br>Fallback]
+    powercap[/sys/class/<br>powercap/]
+    msrdev["/dev/cpu/*/msr"]
+
+    CPUPowerMeter --> raplPowerMeter
+    raplPowerMeter --> powercapReader
+    raplPowerMeter --> msrReader
+    powercapReader --> powercap
+    msrReader --> msrdev
+
+    style CPUPowerMeter fill:#1e88e5,color:#fff
+    style raplPowerMeter fill:#43a047,color:#fff
+    style powercapReader fill:#fb8c00,color:#fff
+    style msrReader fill:#e53935,color:#fff
+    style powercap fill:#fdd835,color:#000
+    style msrdev fill:#8e24aa,color:#fff
+```
+
+## Implementation
+
+### 1. Create abstraction
+
+```go
+type powerReader interface {
+    ReadEnergy(zone EnergyZone) (Energy, error)
+    Available() bool
+    Init() error
+}
+```
+
+### 2. Refactor existing code
+
+* Extract current powercap logic → `powercapReader`
+* Add `msrReader` for `/dev/cpu/*/msr` access
+* Auto-detect in `raplPowerMeter.Init()`
+
+### 3. MSR registers
+
+* PKG: 0x611 (package energy)
+* PP0: 0x639 (core energy)
+* DRAM: 0x619 (memory energy)
+* UNIT: 0x606 (energy units)
+
+## Configuration
+
+```yaml
+device:
+  msr:
+    enabled: false  # opt-in due to security
+    force: false    # force MSR even if powercap available (testing)
+    devicePath: "/dev/cpu/%d/msr"  # MSR device path template
```

Comment:
```suggestion
    path: "/dev/cpu/"  # MSR device path
```

Reply:
Thats right .. the mount path would be `/dev/cpu` but the msr path is `%d/msr`. But you are right we should have host's dev/cpu mounted as `host/dev/cpu`.